testnet/stacks-node/src/burnchains/bitcoin_regtest_controller.rs

```diff
@@ -207,28 +208,40 @@ impl BitcoinRegtestController {
         rest
     }
 
-    fn receive_blocks(&mut self, sync: bool) -> Result<BurnchainTip, BurnchainControllerError> {
+    fn receive_blocks(&mut self, num_burn_blocks_opt: Option<u64>) -> Result<(BurnchainTip, u64), BurnchainControllerError> {
```

Comment:
The prior synchrony behavior was (somewhat) necessary -- when the node first starts up, it should just get started downloading blocks and not wait for the sortition db to process them (because the coordinator wouldn't have started yet). Afterwards, when calling `sync`, it should always wait for the sortition db to be processing (or, perhaps, check the first_burn_block_height) -- with this change, the code only waits on the sortition db if a sortition has already been processed, which introduces a corner case in the event that blocks have all started downloading, but the coordinator hasn't yet processed any of them into the sortition_db. The `sync` method would get called, and return before the sortitions were processed.

Reply:
> The prior synchrony behavior was (somewhat) necessary -- when the node first starts up, it should just get started downloading blocks and not wait for the sortition db to process them (because the coordinator wouldn't have started yet). 

Should it start the coordinator immediately, then?  I think it should, because come mainnet, we'll be starting the coordinator before the first first burnchain block is processed anyway.

I can make it so the node downloads the burnchain blocks for reward cycles N+1+k concurrently with processing sortitions for reward cycle N.
 
Also, I'm having a hard time parsing this:

> Afterwards, when calling sync, it should always wait for the sortition db to have processed the block height it reports (or, perhaps, check the first_burn_block_height) -- with this change, the code only waits on the sortition db if a sortition has already been processed, which introduces a corner case in the event that blocks have all started downloading, but the coordinator hasn't yet processed any of them into the sortition_db.

Can you unpack this a bit more?

Reply:
> Should it start the coordinator immediately, then? I think it should, because come mainnet, we'll be starting the coordinator before the first first burnchain block is processed anyway.

It could -- there's a slight bootstrapping problem I think though, where the coordinator needs to be initialized knowing what the first burn block hash was. That could be solved in a different way than having this struct have `start` and `sync` behave differently, but that's I think the reason that it's structured this way.

> Can you unpack this a bit more?

Sure -- the `sync()` method returns to the run loop once the sortition_db has processed up to the burnchain block tip (or, with this change, up to `num_burn_blocks_opt`). The `wait_for_sortitions` function is what makes that happens. With the changes in this PR, it's possible that function isn't called (if the sortition_db hasn't processed any blocks yet).

Reply:
> Sure -- the sync() method returns to the run loop once the sortition_db has processed up to the burnchain block tip (or, with this change, up to num_burn_blocks_opt). The wait_for_sortitions function is what makes that happens. With the changes in this PR, it's possible that function isn't called (if the sortition_db hasn't processed any blocks yet).

So, the `start()` and `sync()` methods in `BitcoinRegtestController` both call `receive_blocks()`, which will repeatedly call `burnchain.sync_with_indexer()` until it succeeds with `Ok(x)`.  When it succeeds with a new chain tip (`x`), the following code will run:

```
                    let sort_tip = SortitionDB::get_canonical_burn_chain_tip(self.sortdb_ref().conn())
                        .expect("Sortition DB error.");

                    if sort_tip.block_height > 0 {
                        self.wait_for_sortitions(Some(x.block_height));
                    }
```

Isn't this what you're describing?  If the canonical burn chain tip is _not_ the genesis block, then we wait until the chains coordinator has processed all sortitions up to the chain tip returned by `burnchain.sync_with_indexer()`.

Also, per this comment:

> It could -- there's a slight bootstrapping problem I think though, where the coordinator needs to be initialized knowing what the first burn block hash was. That could be solved in a different way than having this struct have start and sync behave differently, but that's I think the reason that it's structured this way.

The chains coordinator gets started up in the `start()` method in `RunLoop` in `neon_node.rs`, per the following:

```
        let (coordinator_receivers, coordinator_senders) = self.coordinator_channels.take()
            .expect("Run loop already started, can only start once after initialization.");

        // Initialize and start the burnchain.
        let mut burnchain = BitcoinRegtestController::new(self.config.clone(), Some(coordinator_senders.clone()));

        /* ... snip ... */

        match burnchain.start() {
            Ok(_) => {},
            Err(e) => {
                warn!("Burnchain controller stopped: {}", e);
                return;
            }
        };

        /* ... snip ... */

        thread::spawn(move || {
            ChainsCoordinator::run(&chainstate_path, burnchain_config, mainnet, chainid,
                                   Some(initial_balances),
                                   block_limit, &coordinator_dispatcher,
                                   coordinator_receivers, |_| {});
        });        
        
        let mut burnchain_tip = burnchain.wait_for_sortitions(None);
```

After this, the code proceeds to call `burnchain.sync()` periodically (with my new changes, it should only call this once Stacks block-processing has slowed down for the past reward cycle).  Doesn't this satisfy the bootstrapping problem?  The `burnchain.start()` method call will download the first reward cycle's burnchain blocks, so when the chains coordinator starts, it will _definitely_ have the first burnchain block hash available to be processed.  The runloop does not proceed to call `burnchain.sync()` again until the chains coordinator has processed that first reward cycle's sortitions.

I can tweak this slightly so the first call to `burnchain.start()` just fetches one burnchain block, which I think satisfies the bootstrapping problem?

Reply:
> Isn't this what you're describing? If the canonical burn chain tip is not the genesis block, then we wait until the chains coordinator has processed all sortitions up to the chain tip returned by burnchain.sync_with_indexer().

Almost -- I think we may want to rename that function. It's returning the current sortition tip, not the burnchain tip. But per your comment on solving the bootstrap problem, yes, the current order of: 1. `start()`, 2. coordinator start, 3. `sync()` solves the bootstrapping problem, but it introduces a race condition between the coordinator thread and the following call to `sync()` -- if, for example, the coordinator thread was never scheduled, the subsequent call to `sync()` will:

1. Fetch some more burnchain blocks into the burnchain db
2. Check the sortition_db's tip, see that it is still at height = 0
3. Return without waiting for the sortition_db to progress

This is problematic, because the neon_node assumes that the sortition_db has processed up to whatever burn block height is returned from the sync() call. This is also why the code previously passed an explicit "sync" parameter to indicate that any call after `start()` should always wait on the sortition_db.

Reply:
Just an update on this:

> I can make it so the node downloads the burnchain blocks for reward cycles N+1+k concurrently with processing sortitions for reward cycle N.

This is actually not something that's easily done with the current architecture -- there's no real way to tell the chains coordinator to _stop_ processing sortitions after a given block height.  The only way to control this is to simply deny it new burnchain blocks to process (which is what the code currently does).  I think making burnchain block downloads concurrent with sortition-processing should be a separate PR.

Reply:
>  The only way to control this is to simply deny it new burnchain blocks to process (which is what the code currently does).

Yep, makes sense, that is what I thought you were proposing before. Yes, I think if we want the coordinator to actually handle this "delayed processing" we should do it in a separate PR.

Reply:
I think this should be addressed now.  The inner `receive_blocks()` method will, at the caller's discretion, call `wait_for_sortitions()`.  The `start()` and `sync()` methods will pass the appropriate value to indicate whether or not `wait_for_sortitions()` gets called, as before.